pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -246,13 +251,18 @@ public void removeSegment(IndexSegment segment) {
     LOGGER.info("Removing upsert metadata for segment: {}", segmentName);
 
     if (!Objects.requireNonNull(segment.getValidDocIds()).getMutableRoaringBitmap().isEmpty()) {
-      // Remove all the record locations that point to the removed segment
-      _primaryKeyToRecordLocationMap.forEach((primaryKey, recordLocation) -> {
-        if (recordLocation.getSegment() == segment) {
-          // Check and remove to prevent removing the key that is just updated
-          _primaryKeyToRecordLocationMap.remove(primaryKey, recordLocation);
+      PeekableIntIterator iterator = segment.getValidDocIds().getMutableRoaringBitmap().getIntIterator();
+      while (iterator.hasNext()) {
+        _reuse.clear();
+        int docId = iterator.next();
+        GenericRow record = segment.getRecord(docId, _reuse);
+        PrimaryKey primaryKey = record.getPrimaryKey(_primaryKeyColumns);
+        if (_primaryKeyToRecordLocationMap.containsKey(primaryKey)
+            && _primaryKeyToRecordLocationMap.get(primaryKey).getSegment() == segment
+            && _primaryKeyToRecordLocationMap.get(primaryKey).getDocId() == docId) {
```

Comment:
No need to check `docId` because we want to remove the entire segment, so as long as the segment matches, we want to remove the entry